persist/dynamo/persister.go

```diff
@@ -0,0 +1,199 @@
+type Persister struct {
+	svc       *dynamodb.DynamoDB
+	tableName *string
+}
+
+// NewPersister constructs a new Persister
+func NewPersister(session *session.Session, tableName string) (Persister, error) {
+	svc := dynamodb.New(session)
+	err := ensureTableExists(svc, tableName)
+	p := Persister{
+		svc:       svc,
+		tableName: &tableName,
+	}
+	return p, err
+}
+
+// getSequenceValue returns a unique sequence value
+func (p *Persister) getSequenceValue() (uint32, error) {
+	uii := &dynamodb.UpdateItemInput{
+		TableName: p.tableName,
+		Key: map[string]*dynamodb.AttributeValue{
+			pkName: {N: aws.String("1")},
+			skName: {S: aws.String("sequence")},
+		},
+		UpdateExpression: aws.String("ADD sequenceValue :i"),
+		ExpressionAttributeValues: map[string]*dynamodb.AttributeValue{
+			":i": {N: aws.String("1")},
+		},
+		ReturnValues: aws.String("ALL_NEW"),
+	}
+	uio, err := p.svc.UpdateItem(uii)
+	if err != nil {
+		return 0, err
+	}
+	i, err := strconv.ParseUint(*uio.Attributes["sequenceValue"].N, 10, 32)
+	return uint32(i), err
+}
+
+func ensureTableExists(svc *dynamodb.DynamoDB, tableName string) error {
+	// See if the table exists already
+	err := waitForTable(svc, tableName)
+	if err == nil {
+		// Perhaps we should compare the schema to what we expect
+		// if so, see https://github.com/dollarshaveclub/dynamo-drift
+		return nil
+	}
+	if aerr, ok := err.(awserr.Error); ok {
+		if aerr.Code() != dynamodb.ErrCodeResourceNotFoundException {
+			return fmt.Errorf("[ERROR] Unexpected error when checking DynamoDB table: %v", err)
+		}
+	} else {
+		return fmt.Errorf("[ERROR] Unexpected error when checking DynamoDB table: %v", err)
+	}
+	log.Printf("[INFO] Creating table %s", tableName)
+	// Couldn't describe it, so try to create it
+	cti := &dynamodb.CreateTableInput{
+		TableName: &tableName,
+		AttributeDefinitions: []*dynamodb.AttributeDefinition{
+			{
+				AttributeName: aws.String(pkName),
+				AttributeType: aws.String("N"),
+			},
+			{
+				AttributeName: aws.String(skName),
+				AttributeType: aws.String("S"),
+			},
+			{
+				AttributeName: aws.String(gsiSkName),
+				AttributeType: aws.String("S"),
+			},
+		},
+		KeySchema: []*dynamodb.KeySchemaElement{
+			{
+				AttributeName: aws.String(pkName),
+				KeyType:       aws.String("HASH"),
+			},
+			{
+				AttributeName: aws.String(skName),
+				KeyType:       aws.String("RANGE"),
+			},
+		},
+		GlobalSecondaryIndexes: []*dynamodb.GlobalSecondaryIndex{
+			{
+				IndexName: aws.String(gsiName),
+				KeySchema: []*dynamodb.KeySchemaElement{
+					{
+						AttributeName: aws.String(skName),
+						KeyType:       aws.String("HASH"),
+					},
+					{
+						AttributeName: aws.String(gsiSkName),
+						KeyType:       aws.String("RANGE"),
+					},
+				},
+				Projection: &dynamodb.Projection{
+					ProjectionType: aws.String("ALL"),
+				},
+				// TODO: How should these values be set/modified?
+				ProvisionedThroughput: &dynamodb.ProvisionedThroughput{
```

Comment:
Can they be set in CloudFormation? Could the table be created in CloudFormation?

Reply:
Yes, it can be done in CloudFormation. But that won't work for tests using DynamoDB Local. I guess I could have CF create the table in a real deploy, then this code would only be used in test environments. Just have to keep the table definitions in sync.

Reply:
Makes sense. It's certainly not urgent.

Reply:
It's easy, I'll go ahead and do it.